File: dotnet/src/Microsoft.AutoGen/AgentChat/GroupChat/GroupChatOptions.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// GroupChatOptions.cs

using Microsoft.AutoGen.AgentChat.Abstractions;

namespace Microsoft.AutoGen.AgentChat.GroupChat;

public struct GroupParticipant(string topicType, string description)
{
    public string TopicType { get; } = topicType;
    public string Description { get; } = description;

    // Destructuring from a tuple
    public GroupParticipant((string topicType, string description) tuple) : this(tuple.topicType, tuple.description)
    {
    }

    // Destructuring to a tuple
    public void Deconstruct(out string topicType, out string description)
    {
        topicType = this.TopicType;
        description = this.Description;
    }

    public static implicit operator GroupParticipant((string topicType, string description) tuple) => new GroupParticipant(tuple);
    public static implicit operator (string topicType, string description)(GroupParticipant participant) => (participant.TopicType, participant.Description);
}

public class GroupChatOptions(string groupTopicType, string outputTopicType)
{
    public string GroupChatTopicType { get; } = groupTopicType;
    public string OutputTopicType { get; } = outputTopicType;

    public ITerminationCondition? TerminationCondition { get; set; }
    public int? MaxTurns { get; set; }

    public Dictionary<string, GroupParticipant> Participants { get; } = new Dictionary<string, GroupParticipant>();
}

File: typescript/src/agentchat/groupchat/GroupChatOptions.ts
import { ITerminationCondition } from "../abstractions/Termination";

/**
 * Represents information about a participant in a group chat.
 */
export class GroupParticipant {
  /**
   * Creates a new group participant.
   * @param topicType The topic type for this participant
   * @param description Description of the participant's capabilities
   */
  constructor(
    public readonly topicType: string,
    public readonly description: string
  ) {}
}

/**
 * Configuration options for a group chat.
 */
export class GroupChatOptions {
  /**
   * Gets the topic type for group chat messages.
   */
  public readonly groupChatTopicType: string;

  /**
   * Gets the topic type for output messages.
   */
  public readonly outputTopicType: string;

  /**
   * Gets or sets the termination condition for the chat.
   */
  public terminationCondition?: ITerminationCondition;

  /**
   * Gets or sets the maximum number of chat turns.
   */
  public maxTurns?: number;

  /**
   * Gets the participants in the chat.
   */
  public readonly participants: Map<string, GroupParticipant> = new Map();

  /**
   * Creates a new instance of GroupChatOptions.
   * @param groupChatTopicType Topic type for group chat messages
   * @param outputTopicType Topic type for output messages
   */
  constructor(groupChatTopicType: string, outputTopicType: string) {
    this.groupChatTopicType = groupChatTopicType;
    this.outputTopicType = outputTopicType;
  }
}
